Allow schema cache path to be defined in the database configuration file

kytrinyx opened the pull request:
### Summary

This updates the database tasks for dumping the Active Record schema cache as well as clearing the schema cache file, allowing the path to be defined in the database configuration YAML file.

As before, the value can also be defined in an `ENV` variable, though this would not work for a multi-db application. If the value is specified neither in the DB config, nor in the ENV, then the path will continue to be derived from the DB config `spec_name`.

### Other Information

In order to make the schema cache database tasks more uniform, I added a `clear_schema_cache` method to the database tasks, in symmetry with the `dump_schema_cache` method. ~Then I pushed the filename logic into those methods, making it easier to test in a way that is understandable.~ _Update: had to back that out, as it was a breaking API change_

The schema cache dumping task was only very lightly tested, so I fleshed those tests out a bit, and the clearing task was not tested, so I added tests for those.

I didn't find any in-depth documentation for the schema cache or the database configuration in the Rails guides, so I only added a bit of method documentation on the new `schema_cache_path` accessor on the database config. If you can think of any other documentation that needs updating, please let me know!

Commit:
Allow schema cache path to be defined in the config file

This updates the database tasks for dumping the Active Record schema cache as
well as clearing the schema cache file, allowing the path to be defined in the
database configuration YAML file.

As before, the value can also be defined in an ENV variable, though this would
not work for a multi-db application. If the value is specified neither in the
DB config, nor in the ENV, then the path will continue to be derived from the
DB config spec_name.

Note that in order to make this change cleaner I also moved a bit of logic
out of a rake task and into the DatabaseTasks class, for symmetry.

We have two rake tasks for the schema cache:

    $ rake db:schema:cache:dump
    $ rake db:schema:cache:clear

The cache:dump task was implemented in DatabaseTasks, but the
cache:clear one was not.

I also added some tests for the behavior that I was changing, since some of
the code paths weren't tested.

tenderlove reviewed (approved):
I like that this allows us to specify the schema file in the db config.  I think this is a güd change.  (@kytrinyx didn't mention it, but we already do this at GitHub, so this is kind of upstreaming our changes.)

rafaelfranca commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -448,12 +442,18 @@ def load_seed
       # ==== Examples:
       #   ActiveRecord::Tasks::DatabaseTasks.dump_schema_cache(ActiveRecord::Base.connection, "tmp/schema_dump.yaml")
-      def dump_schema_cache(conn, filename)
+      def dump_schema_cache(conn, namespace, schema_cache_path:)
```
This change is breaking the API of a public class. I'm fine with it but we need to point in the changelog and also update the example above.

kytrinyx replied:
Ah! I think we should give it a default of `nil`, that way it is compatible with the public API.

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Allow schema cache path to be defined in the database configuration file.
```
Can you expand this entry with an example?

kytrinyx replied:
Good idea! Will do.

rafaelfranca reviewed (commented):
Nice work! We were going to apply a similar change as well. I made two comments about the CHANGELOG.

eileencodes reviewed (approved):
This looks great, can you squash your commits into one?

eileencodes commented on activerecord/lib/active_record/database_configurations/hash_config.rb:
```diff
@@ -87,6 +87,13 @@ def idle_timeout
+      # default will be derived.
+      def schema_cache_path
+        configuration_hash[:schema_cache_path]
+      end
```
Oh I just realized there's one part missing. Can you add a not implemented version of this method in https://github.com/rails/rails/blob/ab3c34e978ca5843cf936d1ea298d543a34df584/activerecord/lib/active_record/database_configurations/database_config.rb?

eileencodes reviewed (changes requested):
Oops I missed something when I reviewed before

seejohnrun commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -409,14 +409,8 @@ def dump_filename(namespace, format = ActiveRecord::Base.schema_format)
-        end
-
-        ENV["SCHEMA_CACHE"] || File.join(ActiveRecord::Tasks::DatabaseTasks.db_dir, filename)
+      def cache_dump_filename(namespace, schema_cache_path = nil)
```
What do you think of renaming `namespace` in here to `spec_name` for consistency with other places we're using this value?

kytrinyx replied:
Yeah, good call.

kytrinyx replied:
Hm. In the context of the filename, I'm not actually sure I want to call it `spec_name`. The `spec_name` is specific to the database config, and here we don't see that it came from the db config, so `spec_name` would be more confusing unless you happen to know off the top of your head where this is coming from.

seejohnrun commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -409,14 +409,8 @@ def dump_filename(namespace, format = ActiveRecord::Base.schema_format)
-
-        ENV["SCHEMA_CACHE"] || File.join(ActiveRecord::Tasks::DatabaseTasks.db_dir, filename)
+      def cache_dump_filename(namespace, schema_cache_path = nil)
+        schema_cache_path || ENV["SCHEMA_CACHE"] || default_schema_cache_path(namespace)
```
Most of the other methods in `HashConfig` contain their own default values. We could do that here by having the method in the config default to a value based on `spec_name`. Then this line would be something like:

`ENV["SCHEMA_CACHE"] || schema_cache_path`

and we could skip passing `namespace / spec_name` into this method. Alternatively, we might just pass `db_config` into this method and take the value we need of it. Just an idea!

kytrinyx replied:
I like the idea of having the default in the database config. That's much cleaner!

The db config should still not know about ENV, though, right?

kytrinyx replied:
Actually, I changed my mind :(

In order to come up with the derived value we need to know the default database directory path, which we would have to get from `Rails.application.config.paths["db"].first` or some such thing. I don't think the database config should reference the top-level Rails config.

kytrinyx replied:
> Alternatively, we might just pass db_config into this method and take the value we need of it.

I'd prefer not to, since that passes a lot more into the method than it needs.

Overall I'm a bit uncomfortable with the fact that this method needs things from so many places in order to do its job, so it doesn't really _belong_ anywhere!

- `Rails.application.config.paths["db"].first`
- `db_config.spec_name`
- `db_config.schema_cache_path`
- `ENV["SCHEMA_CACHE"]

That's a lot of knowledge.

seejohnrun reviewed (commented):
This PR looks good! I left a few comments around consistency with other `HashConfig` / `DatabaseConfig` method 🎉

rafaelfranca commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -448,12 +442,18 @@ def load_seed
       # ==== Examples:
       #   ActiveRecord::Tasks::DatabaseTasks.dump_schema_cache(ActiveRecord::Base.connection, "tmp/schema_dump.yaml")
-      def dump_schema_cache(conn, filename)
+      def dump_schema_cache(conn, namespace, schema_cache_path: nil)
```
This still has a different API, before it received a filename, now it is a namespace. I think we need to update the example above and also add a CHANGELOG entry since `ActiveRecord::Tasks::DatabaseTasks.dump_schema_cache(ActiveRecord::Base.connection, "tmp/schema_dump.yaml")` will not work anymore.

kytrinyx replied:
Oh! I completely missed that. Thank you so much for the careful review.

I am going to revert this API back to what it was. I think from @seejohnrun's comments I see a better way of doing it.

kytrinyx replied:
Hm, ok looking at this more, I will need to change the API of the `ActiveRecord::Tasks::DatabaseTasks.cache_dump_filename` method.

The cache_dump_filename doesn't have any explicit documentation, but it's not :nodoc: or private either, so we definitely will need something in the changelog for that.

kytrinyx replied:
Wait. No I won't. What am I thinking? The `cache_dump_filename` method already takes the namespace.

kytrinyx commented:
I've reworked this based on the discussions so far:

1. This reverts back to the original API of the `dump_schema_cache` method
2. It changes the API of the `schema_dump_filename` but it is backwards compatible
3. I've added the method with a `raise NotImplementedError` in the `DatabaseConfig` class
4. I decided not to change the parameter to `schema_dump_filename` to `spec_name` but rather keep the original, since we're no longer in a context of a `DatabaseConfig` object.
5. I tried moving the default name implementation into the `HashConfig#schema_cache_path`, but realized that it was too gross for the db config class to know about the rails top-level config, so I have left it in the `DatabaseTasks`.

Let me know if there are any other changes you can think of, and if not I'll squash all the commits.
